File: mobile-sim/python/mobilesim/rosie/perception/WorldObjectManager.py
from pysoarlib import *

from .WorldObject import WorldObject
from string import digits

class WorldObjectManager:
    def __init__(self):
        self.objects = {}
        self.objs_to_remove = {}

        self.objects_id = None
        self.wm_dirty = False

    def get_object(self, handle):
        return self.objects.get(handle, None)

    def get_object_by_cat(self, cat):
        for obj in self.objects.values():
            if cat in obj.get_property("category"):
                return obj
        return None

    def get_object_container(self, obj):
        if obj is None:
            return None
        for container in self.objects.values():
            if obj.id in container.contents:
                return container
        return None

    def update(self, new_object_data):
        new_ids = set( od.id for od in new_object_data.objects )
        for obj_data in new_object_data.objects:
            obj_id = obj_data.id

            obj = self.objects.get(obj_id, None)
            if obj is None:
                if obj in self.objs_to_remove:
                    # Object was going to be removed
                    # Transfer it to the normal list and update
                    obj = self.objs_to_remove[obj_id]
                    del self.objs_to_remove[obj_id]
                else:
                    # It's a new object, add it to the map
                    obj = WorldObject(obj_data);
                self.objects[obj_id] = obj

            obj.update(obj_data)

        old_ids = [ obj_id for obj_id in self.objects if obj_id not in new_ids ]
        for old_id in old_ids:
            self.objs_to_remove[old_id] = self.objects[old_id]
            del self.objects[old_id]

        self.wm_dirty = True


    #### METHODS TO UPDATE WORKING MEMORY ####

    def is_added(self):
        return self.objects_id is not None

    def add_to_wm(self, parent_id, svs_commands):
        if self.objects_id is not None:
            self.update_wm(parent_id, svs_commands)
            return

        self.objects_id = parent_id.CreateIdWME("objects")
        for obj in self.objects.values():
            obj.update_wm(self.objects_id, svs_commands)

        self.wm_dirty = False

    def update_wm(self, parent_id, svs_commands):
        if self.objects_id is None:
            self.add_to_wm(parent_id, svs_commands)
            return

        for obj in self.objects.values():
            obj.update_wm(self.objects_id, svs_commands)

        for obj in self.objs_to_remove.values():
            obj.remove_from_wm(svs_commands)
        self.objs_to_remove.clear()

        self.wm_dirty = False

    def remove_from_wm(self, svs_commands):
        if self.objects_id is None: return

        for obj in self.objects.values():
            obj.remove_from_wm(svs_commands)

        for obj in self.objs_to_remove.values():
            obj.remove_from_wm(svs_commands)
        self.objs_to_remove.clear()

        self.objects_id.DestroyWME()
        self.objects_id = None

File: mobile-sim/java/src/soargroup/rosie/perception/WorldObjectManager.java
package soargroup.rosie.perception;

import java.io.IOException;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;

import edu.umich.rosie.soar.SoarUtil;
import edu.umich.rosie.soar.ISVSObject;

import sml.Identifier;

// LCM Types
import lcm.lcm.*;
import soargroup.mobilesim.lcmtypes.object_data_t;
import soargroup.mobilesim.lcmtypes.object_data_list_t;

public class WorldObjectManager implements ISVSObject {
	private HashMap<Integer, WorldObject> objects;
	private HashMap<Integer, WorldObject> objsToRemove;
	private boolean gotUpdate = true;

    private Identifier objectsId = null;

    public WorldObjectManager(){
    	objects = new HashMap<Integer, WorldObject>();
    	objsToRemove = new HashMap<Integer, WorldObject>();
    }

	public WorldObject getObject(Integer id){
		return objects.get(id);
	}

	public void update(object_data_list_t newObjectData){
    	// Set of objects that didn't appear in the new update
    	// (remove ids as we see them)
    	HashSet<Integer> oldIds = new HashSet<Integer>();
    	oldIds.addAll(objects.keySet());

		for(object_data_t objData : newObjectData.objects){
			if(!objData.visible){
				// Ignore objects that aren't visible
				continue;
			}
			Integer objID = objData.id;
			if(oldIds.contains(objID)){
				oldIds.remove(objID);
			}

			WorldObject obj = objects.get(objID);
			if(obj == null){
				if(objsToRemove.containsKey(objID)){
    				// Object was going to be removed
    				// Transfer it to the normal list and update
					obj = objsToRemove.get(objID);
					objsToRemove.remove(objID);
				} else {
    				// It's a new object, add it to the map
					obj = new WorldObject(objData);
					objects.put(objID, obj);
				}
			}

    		obj.update(objData);
    	}

    	for(Integer oldID : oldIds){
    		WorldObject oldObj = objects.get(oldID);
    		objects.remove(oldID);
    		objsToRemove.put(oldID, oldObj);
    	}

		gotUpdate = true;
    }

	// Returns true if the object is currently in working memory
	public boolean isAdded(){
		return (objectsId != null);
	}

	// Add the object to working memory, on the given identifier
	public void addToWM(Identifier parentId, StringBuilder svsCommands){
		if (objectsId != null){ return; }

		objectsId = parentId.CreateIdWME("objects");
		for(WorldObject obj : objects.values()){
			obj.addToWM(objectsId, svsCommands);
		}
	}
	
	// Update is called before input phase, changes can be made to WM
	public void updateWM(StringBuilder svsCommands){
		if (objectsId == null){ return; }
		if (!gotUpdate){ return; }

		for(WorldObject obj : objects.values()){
			if(obj.isAdded()){
				obj.updateWM(svsCommands);
			} else {
				obj.addToWM(objectsId, svsCommands);
			}
		}
    	for(WorldObject obj : objsToRemove.values()){
    		obj.removeFromWM(svsCommands);
    	}
    	objsToRemove.clear();
		gotUpdate = false;
	}
	
	// Release all references to SML objects and remove from WM
	public void removeFromWM(StringBuilder svsCommands){
		if (objectsId == null){ return; }

		for(WorldObject obj : objects.values()){
    		obj.removeFromWM(svsCommands);
		}
    	for(WorldObject obj : objsToRemove.values()){
    		obj.removeFromWM(svsCommands);
    	}
		objsToRemove.clear();

		objectsId.DestroyWME();
		objectsId = null;
	}
}
